File: server/modules/auth/services/password-policy.ts
/**
 * Força mínima de senha — regra compartilhada entre cadastro e troca de senha de perfil.
 *
 * Migrado de legacy/backend/models/profilePasswordPolicy.ts (completo agora que
 * `modules/profile/` — dono de `validateProfileNewPasswordStrength` — migrou).
 */
import { authWorkerBcrypt } from './auth-worker-client.js';
import { rustValidatePasswordStrength } from './auth-rust-bridge.js';
const COMMON_WEAK_PASSWORDS = new Set(
  [
    'password',
    '12345678',
    '123456789',
    'qwerty123',
    'genesis',
    'genesisminer',
    'welcome1',
    'senha123',
    'palavrapasse',
    'abc123456'
  ].map((s) => s.toLowerCase())
);

const PASSWORD_MIN_LENGTH = 6;
const PASSWORD_STRENGTH_MAX_LENGTH = 50;

export type ProfilePasswordStrength = { ok: true } | { ok: false; error: string };

export function validatePasswordStrengthPolicy(newPassword: string): ProfilePasswordStrength {
  const rust = rustValidatePasswordStrength(String(newPassword || ''));
  if (rust) return rust;
  const p = String(newPassword || '');
  if (p.length < PASSWORD_MIN_LENGTH) {
    return { ok: false, error: `Password must be at least ${PASSWORD_MIN_LENGTH} characters.` };
  }
  if (p.length > PASSWORD_STRENGTH_MAX_LENGTH) {
    return { ok: false, error: 'The new password is too long.' };
  }
  if (COMMON_WEAK_PASSWORDS.has(p.toLowerCase())) {
    return { ok: false, error: 'This password is too common. Choose another.' };
  }
  return { ok: true };
}

/** Regra de troca de senha autenticada: além da força mínima, não pode repetir a senha atual. */
export async function validateProfileNewPasswordStrength(
  newPassword: string,
  currentHash: string | null | undefined
): Promise<ProfilePasswordStrength> {
  const base = validatePasswordStrengthPolicy(newPassword);
  if (!base.ok) return base;
  const p = String(newPassword || '');
  if (currentHash && (await authWorkerBcrypt.compare(p, currentHash))) {
    return { ok: false, error: 'The new password cannot be the same as the current one.' };
  }
  return { ok: true };
}

File: rust/genesis-core/src/auth/password_policy.rs
use super::constants::{COMMON_WEAK_PASSWORDS, PASSWORD_MIN_LENGTH, PASSWORD_STRENGTH_MAX_LENGTH};
use super::types::PolicyResult;

pub fn validate_password_strength_policy(new_password: &str) -> PolicyResult {
    let p = new_password;
    if p.len() < PASSWORD_MIN_LENGTH {
        return PolicyResult::err(format!(
            "Password must be at least {PASSWORD_MIN_LENGTH} characters."
        ));
    }
    if p.len() > PASSWORD_STRENGTH_MAX_LENGTH {
        return PolicyResult::err("The new password is too long.");
    }
    let lower = p.to_lowercase();
    if COMMON_WEAK_PASSWORDS.contains(&lower.as_str()) {
        return PolicyResult::err("This password is too common. Choose another.");
    }
    PolicyResult::ok()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_short() {
        assert!(!validate_password_strength_policy("12345").is_ok());
    }

    #[test]
    fn rejects_weak() {
        assert!(!validate_password_strength_policy("password").is_ok());
    }

    #[test]
    fn accepts_reasonable() {
        assert!(validate_password_strength_policy("hunter2x").is_ok());
    }
}
